test/common/test_context.go

```diff
@@ -80,76 +127,71 @@ func TestEnv(additionalFlagFuncs ...func(set *pflag.FlagSet)) env.Environment {
 
 	additionalFlagFuncs = append(additionalFlagFuncs, f)
 
-	// will be used as default value and overridden if --file.location={{location}} is passed to go test
-	additionalFlagFuncs = append(additionalFlagFuncs, SetTestFileLocation)
-
 	return sm.DefaultEnv(additionalFlagFuncs...)
 }
 
-type ContextParams struct {
-	RegisterExtensions func(smb *sm.ServiceManagerBuilder)
-	DefaultTokenClaims map[string]interface{}
-	Env                env.Environment
+func (tcb *TestContextBuilder) WithDefaultEnv(envCreateFunc func(f ...func(set *pflag.FlagSet)) env.Environment) *TestContextBuilder {
+	tcb.Environment = envCreateFunc
+
+	return tcb
 }
 
-func NewSMServer(params *ContextParams, issuerURL string) *httptest.Server {
-	var smEnv env.Environment
-	if params.Env != nil {
-		smEnv = params.Env
-	} else {
-		smEnv = e
+func (tcb *TestContextBuilder) WithAdditionalFakeServers(additionalFakeServers map[string]FakeServer) *TestContextBuilder {
+	if tcb.Servers == nil {
+		tcb.Servers = make(map[string]FakeServer, 0)
 	}
 
-	smEnv.Set("api.token_issuer_url", issuerURL)
+	for name, server := range additionalFakeServers {
+		tcb.Servers[name] = server
+	}
 
-	flag.VisitAll(func(flag *flag.Flag) {
-		if flag.Value.String() != "" {
-			// if any of the go test flags have been set, propagate the value in sm env with highest prio
-			// when env exposes the pflagset it would be better to instead override the pflag value instead
-			smEnv.Set(flag.Name, flag.Value.String())
-		}
-	})
+	return tcb
+}
 
-	ctx, cancel := context.WithCancel(context.Background())
-	s := struct {
-		Log *log.Settings
-	}{
-		Log: &log.Settings{
-			Output: ginkgo.GinkgoWriter,
-		},
-	}
-	err := smEnv.Unmarshal(&s)
-	if err != nil {
-		panic(err)
-	}
-	ctx = log.Configure(ctx, s.Log)
-	smanagerBuilder := sm.New(ctx, cancel, smEnv)
-	if params.RegisterExtensions != nil {
-		params.RegisterExtensions(smanagerBuilder)
-	}
-	serviceManager := smanagerBuilder.Build()
-	return httptest.NewServer(serviceManager.Server.Router)
+func (tcb *TestContextBuilder) WithDefaultTokenClaims(defaultTokenClaims map[string]interface{}) *TestContextBuilder {
+	tcb.defaultTokenClaims = defaultTokenClaims
+
+	return tcb
 }
 
-func NewTestContext(params *ContextParams) *TestContext {
-	if params == nil {
-		params = &ContextParams{}
-	}
+func (tcb *TestContextBuilder) WithEnvPreExtensions(fs ...func(set *pflag.FlagSet)) *TestContextBuilder {
+	tcb.envPreHooks = append(tcb.envPreHooks, fs...)
+
+	return tcb
+}
+
+func (tcb *TestContextBuilder) WithEnvPostExtensions(fs ...func(e env.Environment, servers map[string]FakeServer)) *TestContextBuilder {
```

Comment:
Should these post hooks take servers?

Reply:
in post hooks you often want to env.Set something and it could be a url to a server for some client to call (see the complex example)